cpp/src/arrow/compute/kernels/vector_selection.cc

```diff
@@ -365,6 +370,122 @@ struct PrimitiveTakeImpl {
     }
     out_arr->null_count = out_arr->length - valid_count;
   }
+
+  static void Exec(const ChunkedArray& values, const ChunkedArray& indices_chunked,
+                   ArrayDataVector* out_chunks) {
+    auto values_resolver = ChunkResolver(values.chunks());
+    const std::vector<const ValueCType*> values_data = MapVector(
+        [](const auto& x) { return x->data()->template GetValues<ValueCType>(1); },
+        values.chunks());
+    const std::vector<const uint8_t*> values_is_valid =
+        MapVector([](const auto& x) { return x->null_bitmap_data(); }, values.chunks());
+    const std::vector<int64_t> values_offset =
+        MapVector([](const auto& x) { return x->offset(); }, values.chunks());
+
+    for (int i = 0; i < indices_chunked.num_chunks(); ++i) {
+      const std::shared_ptr<Array>& indices_chunk = indices_chunked.chunk(i);
+      ArrayData* out_arr = (*out_chunks)[i].get();
+      ValueCType* out = out_arr->GetMutableValues<ValueCType>(1);
+      uint8_t* out_is_valid = out_arr->buffers[0]->mutable_data();
+      int64_t out_offset = out_arr->offset;
+
+      int64_t valid_count = 0;
+
+      int64_t position = 0;  // Position in output array
+      const ArraySpan indices = ArraySpan(*indices_chunk.get()->data());
+      // TODO: How do we reduce duplication of code?
```

Comment:
See https://github.com/apache/arrow/blob/8474ee5a3ed725d4bb56c75fc1b13a53cba1fd1f/cpp/src/arrow/compute/kernels/vector_sort.cc#L690-L711 and https://github.com/apache/arrow/blob/8474ee5a3ed725d4bb56c75fc1b13a53cba1fd1f/cpp/src/arrow/compute/kernels/vector_sort.cc#L844-L889 for ideas of how to parametrize this.